lang/cs/Org.Apache.REEF.IMRU/OnREEF/StateMachine/SystemStateMachine.cs

```diff
@@ -0,0 +1,102 @@
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.NotRecoverable), SystemState.Fail },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.AllTasksAreRunning), SystemState.TasksRunning },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.FailedEvaluator), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.FailedTask), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.FailedEvaluator), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.FailedTask), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.AllTasksAreCompleted), SystemState.TasksCompleted },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.ShuttingDown, SystemStateEvent.Recover), SystemState.WaitingForEvaluator },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.ShuttingDown, SystemStateEvent.NotRecoverable), SystemState.Fail }
+        });
+
+        private readonly object _stateLock = new object();
+        private SystemState _currentState;
+
+        /// <summary>
+        /// create a new SysteState starting from WaitingForEvaluator
+        /// </summary>
+        internal SystemStateMachine()
+        {
+            _currentState = SystemState.WaitingForEvaluator;
+        }
+
+        /// <summary>
+        /// Returns current state
+        /// </summary>
+        internal SystemState CurrentState
+        {
+            get
+            {
+                lock (_stateLock)
+                {
+                    return _currentState;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Based on the event and current state to determine the next state
+        /// </summary>
+        /// <param name="systemEvent"></param>
+        /// <returns></returns>
+        internal SystemState GetNext(SystemStateEvent systemEvent)
+        {
+            var transition = new StateTransition<SystemState, SystemStateEvent>(_currentState, systemEvent);
+            SystemState nextState;
+            if (!Transitions.TryGetValue(transition, out nextState))
+            {
+                throw new SystemStateTransitionException(_currentState, systemEvent);
+            }
+            return nextState;
+        }
+
+        /// <summary>
+        /// Move to next state based on the event given
+        /// </summary>
+        /// <param name="stateEvent"></param>
+        /// <returns></returns>
+        internal SystemState MoveNext(SystemStateEvent stateEvent)
```

Comment:
So if my understanding is right, only one TaskEvent will be generated at any given time right since these TaskEvents are the synchronization barriers in the code. Why do we need locks here then?